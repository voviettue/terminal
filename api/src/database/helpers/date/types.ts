import { DatabaseHelper } from '../types';
import { parseISO } from 'date-fns';

export abstract class DateHelper extends DatabaseHelper {
	parse(date: string): string {
		return date;
	}
	readTimestampString(date: string): string {
		return date;
	}
	writeTimestamp(date: string): Date {
		return parseISO(date);
	}
	fieldFlagForField(_fieldType: string): string {
		return '';
	}
}
