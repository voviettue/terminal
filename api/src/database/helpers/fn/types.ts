import { Query, SchemaOverview } from '@directus/shared/types';
import { Knex } from 'knex';
import { applyFilter } from '../../../utils/apply-query';
import { DatabaseHelper } from '../types';

export type FnHelperOptions = {
	type?: string;
	query?: Query;
};

export abstract class FnHelper extends DatabaseHelper {
	constructor(protected knex: Knex, protected schema: SchemaOverview) {
		super(knex);
		this.schema = schema;
	}

	abstract year(table: string, column: string, options?: FnHelperOptions): Knex.Raw;
	abstract month(table: string, column: string, options?: FnHelperOptions): Knex.Raw;
	abstract week(table: string, column: string, options?: FnHelperOptions): Knex.Raw;
	abstract day(table: string, column: string, options?: FnHelperOptions): Knex.Raw;
	abstract weekday(table: string, column: string, options?: FnHelperOptions): Knex.Raw;
	abstract hour(table: string, column: string, options?: FnHelperOptions): Knex.Raw;
	abstract minute(table: string, column: string, options?: FnHelperOptions): Knex.Raw;
	abstract second(table: string, column: string, options?: FnHelperOptions): Knex.Raw;
	abstract count(table: string, column: string, options?: FnHelperOptions): Knex.Raw;

	protected _relationalCount(table: string, column: string, options?: FnHelperOptions): Knex.Raw {
		const relation = this.schema.relations.find(
			(relation) => relation.related_collection === table && relation?.meta?.one_field === column
		);

		const currentPrimary = this.schema.collections[table].primary;

		if (!relation) {
			throw new Error(`Field ${table}.${column} isn't a nested relational collection`);
		}

		let countQuery = this.knex
			.count('*')
			.from(relation.collection)
			.where(relation.field, '=', this.knex.raw(`??.??`, [table, currentPrimary]));

		if (options?.query?.filter) {
			countQuery = applyFilter(this.knex, this.schema, countQuery, options.query.filter, relation.collection, false);
		}

		return this.knex.raw('(' + countQuery.toQuery() + ')');
	}
}
