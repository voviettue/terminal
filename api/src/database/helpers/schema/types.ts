import { getDatabaseClient } from '../../index';
import { DatabaseHelper } from '../types';
import { KNEX_TYPES } from '@directus/shared/constants';

type Clients = 'mysql' | 'postgres' | 'cockroachdb' | 'sqlite' | 'oracle' | 'mssql' | 'redshift';

export type Options = { nullable?: boolean; default?: any; length?: number };

export abstract class SchemaHelper extends DatabaseHelper {
	isOneOfClients(clients: Clients[]): boolean {
		return clients.includes(getDatabaseClient(this.knex));
	}

	async changeNullable(table: string, column: string, nullable: boolean): Promise<void> {
		await this.knex.schema.alterTable(table, (builder) => {
			if (nullable) {
				builder.setNullable(column);
			} else {
				builder.dropNullable(column);
			}
		});
	}

	async changeToType(
		table: string,
		column: string,
		type: typeof KNEX_TYPES[number],
		options: Options = {}
	): Promise<void> {
		await this.knex.schema.alterTable(table, (builder) => {
			const b = type === 'string' ? builder.string(column, options.length) : builder[type](column);

			if (options.nullable === true) {
				b.nullable();
			}

			if (options.nullable === false) {
				b.notNullable();
			}

			if (options.default !== undefined) {
				b.defaultTo(options.default);
			}

			b.alter();
		});
	}

	protected async changeToTypeByCopy(
		table: string,
		column: string,
		type: typeof KNEX_TYPES[number],
		options: Options
	): Promise<void> {
		const tempName = `${column}__temp`;

		await this.knex.schema.alterTable(table, (builder) => {
			const col = type === 'string' ? builder.string(tempName, options.length) : builder[type](tempName);

			if (options.default !== undefined) {
				col.defaultTo(options.default);
			}

			// Force new temporary column to be nullable (required, as there will already be rows in
			// the table)
			col.nullable();
		});

		await this.knex(table).update(tempName, this.knex.ref(column));

		await this.knex.schema.alterTable(table, (builder) => {
			builder.dropColumn(column);
		});

		await this.knex.schema.alterTable(table, (builder) => {
			builder.renameColumn(tempName, column);
		});

		// We're altering the temporary column here. That starts nullable, so we only want to set it
		// to NOT NULL when applicable
		if (options.nullable === false) {
			await this.changeNullable(table, column, options.nullable);
		}
	}

	async preColumnChange(): Promise<boolean> {
		return false;
	}

	async postColumnChange(): Promise<void> {
		return;
	}
}
