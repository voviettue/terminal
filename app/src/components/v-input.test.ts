import { test, expect } from 'vitest';
import { mount } from '@vue/test-utils';

import VInput from './v-input.vue';
import { GlobalMountOptions } from '@vue/test-utils/dist/types';
import { Focus } from '@/__utils__/focus';

const global: GlobalMountOptions = {
	stubs: ['v-icon'],
	directives: {
		focus: Focus,
	},
};

test('Mount component', () => {
	expect(VInput).toBeTruthy();

	const wrapper = mount(VInput, {
		global,
	});

	expect(wrapper.html()).toMatchSnapshot();
});

test('modelValue prop', async () => {
	const wrapper = mount(VInput, {
		props: {
			modelValue: 'my value',
		},
		global,
	});

	expect(wrapper.get('input').element.value).toBe('my value');

	await wrapper.find('input').setValue('my value1');

	expect(wrapper.emitted()['update:modelValue'][0]).toEqual(['my value1']);
});

test('modelValue trim', async () => {
	const wrapper = mount(VInput, {
		props: {
			modelValue: '  please trim that beard    ',
			trim: true,
		},
		global,
	});

	await wrapper.find('input').trigger('blur');

	expect(wrapper.emitted()['update:modelValue'][0]).toEqual(['please trim that beard']);
});

test('modelValue dbSafe', async () => {
	const wrapper = mount(VInput, {
		props: {
			modelValue: 'this $hould be DÃŸ save!!',
			dbSafe: true,
		},
		global,
	});

	await wrapper.find('input').trigger('input');

	expect(wrapper.emitted()['update:modelValue'][0]).toEqual(['this_hould_be_D_save']);
});
