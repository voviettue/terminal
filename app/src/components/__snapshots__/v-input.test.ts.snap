// Vitest Snapshot v1

exports[`Mount component 1`] = `
"<div class=\\"v-input full-width\\" data-v-2a853453=\\"\\">
  <!--v-if-->
  <div class=\\"input\\" data-v-2a853453=\\"\\">
    <!--v-if-->
    <!--v-if--><input autocomplete=\\"off\\" type=\\"text\\" step=\\"1\\" data-v-2a853453=\\"\\">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>"
`;
